### ISSUE TITLE
Curious about application, hoping for some clarity!

### ISSUE THREAD

**Issue by edisonywh** [olegantonyan/typerb#5]
Hey, I read through the README and wasn't very sure what's the exact usage of this. Doesn't the code below does the exact same thing?

![carbon_2018-11-24_15-49-58](https://user-images.githubusercontent.com/20277437/48965872-a1cc7f00-f000-11e8-8097-30e6431ec36c.png)

(I monkey patched on `Kernel` which is arguable, but that's not the point of this issue)

A couple of questions in regards to type checking in Ruby:
- Isn't Ruby generally big on **duck typing?** If an argument responds to `.each`, then we don't really care if it's `Array` or something else.
- Isn't the big argument for type safety that it's done during _compile time_ so that the compiler notifies us before we even ship bad code? (I do reckon there's a use case for runtime check, but I'm curious to what you think)

I come in peace! Mainly curious about it and want to have a discussion around it :)

**Comment by olegantonyan** [comment]
> Doesn't the code below does the exact same thing?

Essentially yes. The difference is that I'd like to get the variable name, and call typecheck method on the object itself.

> Isn't Ruby generally big on duck typing? If an argument responds to .each, then we don't really care if it's Array or something else

There couple of reasons why I found these type checks useful. Duck-typing is fun as long as you're dealing with perfect code, where all the developers always check for nils, write meaningful tests, follow SOLID, etc. In the real world big legacy codebase will destroy all the benefits of duck-typing. 

TL;DR:

1. Better readability - you're telling other people what kind of arguments you expect. This is a huge deal in large codebase;
2. Defeat `nil`s;
3. Debug problems by looking at error monitoring system.

Longer version based on true codebase:

```ruby
class SomeClass
  def do_something(data)
    ... 
   # bla-bla-bla 100 lines of ifs-elses, early returns
   do_something_else(data)
   # bla-bla-bla another 100 lines of mess
```

Just look at this method and try to understand what the heck is `data` ? Ok, there should be a test... nope. Ok, I can add `puts data.class` in the beginning and hope that there will be one and only one kind of objects. So, for example I found that `data` is Array here. Ok, is there a guarantee that it can't be something completely different? Nope! It can be anything. This method may be called dynamically (oh, we love meta-programming) from some rare places. It's hard to find all those places. And there might be a type check 5 method down the stack `data.is_a?(Array) ? do_this : do_that`. Ok, let's look at `do_that` method.. you've got the idea. 

```ruby
class SomeClass
  def do_something(data)
    data.type!(Array)
```
This tells me much more about the method. It will also raise an exception which I'll see in the error monitoring system if data was not Array, with stacktrace so I can find exact place where it was called. This saves a lot of time.
